// @metaplex-foundation/mpl-core/dist/src/generated/accounts/groupV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGroupV1GpaBuilder = exports.safeFetchAllGroupV1 = exports.fetchAllGroupV1 = exports.safeFetchGroupV1 = exports.fetchGroupV1 = exports.deserializeGroupV1 = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const types_1 = require("../types");
function deserializeGroupV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, (0, hooked_1.getGroupV1AccountDataSerializer)());
}
exports.deserializeGroupV1 = deserializeGroupV1;
async function fetchGroupV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'GroupV1');
    return deserializeGroupV1(maybeAccount);
}
exports.fetchGroupV1 = fetchGroupV1;
async function safeFetchGroupV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializeGroupV1(maybeAccount) : null;
}
exports.safeFetchGroupV1 = safeFetchGroupV1;
async function fetchAllGroupV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'GroupV1');
        return deserializeGroupV1(maybeAccount);
    });
}
exports.fetchAllGroupV1 = fetchAllGroupV1;
async function safeFetchAllGroupV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializeGroupV1(maybeAccount));
}
exports.safeFetchAllGroupV1 = safeFetchAllGroupV1;
function getGroupV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        updateAuthority: [1, (0, serializers_1.publicKey)()],
        name: [33, (0, serializers_1.string)()],
        uri: [null, (0, serializers_1.string)()],
        collections: [null, (0, serializers_1.array)((0, serializers_1.publicKey)())],
        groups: [null, (0, serializers_1.array)((0, serializers_1.publicKey)())],
        parentGroups: [null, (0, serializers_1.array)((0, serializers_1.publicKey)())],
        assets: [null, (0, serializers_1.array)((0, serializers_1.publicKey)())],
    })
        .deserializeUsing((account) => deserializeGroupV1(account))
        .whereField('key', types_1.Key.GroupV1);
}
exports.getGroupV1GpaBuilder = getGroupV1GpaBuilder;
