// @metaplex-foundation/mpl-core/dist/src/generated/accounts/pluginRegistryV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPluginRegistryV1GpaBuilder = exports.safeFetchAllPluginRegistryV1 = exports.fetchAllPluginRegistryV1 = exports.safeFetchPluginRegistryV1 = exports.fetchPluginRegistryV1 = exports.deserializePluginRegistryV1 = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const hooked_1 = require("../../hooked");
const types_1 = require("../types");
function deserializePluginRegistryV1(rawAccount) {
    return (0, umi_1.deserializeAccount)(rawAccount, (0, hooked_1.getPluginRegistryV1AccountDataSerializer)());
}
exports.deserializePluginRegistryV1 = deserializePluginRegistryV1;
async function fetchPluginRegistryV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    (0, umi_1.assertAccountExists)(maybeAccount, 'PluginRegistryV1');
    return deserializePluginRegistryV1(maybeAccount);
}
exports.fetchPluginRegistryV1 = fetchPluginRegistryV1;
async function safeFetchPluginRegistryV1(context, publicKey, options) {
    const maybeAccount = await context.rpc.getAccount((0, umi_1.publicKey)(publicKey, false), options);
    return maybeAccount.exists ? deserializePluginRegistryV1(maybeAccount) : null;
}
exports.safeFetchPluginRegistryV1 = safeFetchPluginRegistryV1;
async function fetchAllPluginRegistryV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts.map((maybeAccount) => {
        (0, umi_1.assertAccountExists)(maybeAccount, 'PluginRegistryV1');
        return deserializePluginRegistryV1(maybeAccount);
    });
}
exports.fetchAllPluginRegistryV1 = fetchAllPluginRegistryV1;
async function safeFetchAllPluginRegistryV1(context, publicKeys, options) {
    const maybeAccounts = await context.rpc.getAccounts(publicKeys.map((key) => (0, umi_1.publicKey)(key, false)), options);
    return maybeAccounts
        .filter((maybeAccount) => maybeAccount.exists)
        .map((maybeAccount) => deserializePluginRegistryV1(maybeAccount));
}
exports.safeFetchAllPluginRegistryV1 = safeFetchAllPluginRegistryV1;
function getPluginRegistryV1GpaBuilder(context) {
    const programId = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
    return (0, umi_1.gpaBuilder)(context, programId)
        .registerFields({
        key: [0, (0, types_1.getKeySerializer)()],
        registry: [1, (0, serializers_1.array)((0, types_1.getRegistryRecordSerializer)())],
        externalRegistry: [null, (0, serializers_1.array)((0, types_1.getExternalRegistryRecordSerializer)())],
    })
        .deserializeUsing((account) => deserializePluginRegistryV1(account));
}
exports.getPluginRegistryV1GpaBuilder = getPluginRegistryV1GpaBuilder;
