// @metaplex-foundation/mpl-core/dist/src/hooked/groupAccountData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getGroupV1AccountDataSerializer = void 0;
const generated_1 = require("../generated");
const groupV1AccountData_1 = require("../generated/types/groupV1AccountData");
const plugins_1 = require("../plugins");
const externalPluginAdapters_1 = require("../plugins/externalPluginAdapters");
const pluginRegistryV1Data_1 = require("./pluginRegistryV1Data");
const getGroupV1AccountDataSerializer = () => ({
    description: 'GroupAccountData',
    fixedSize: null,
    maxSize: null,
    serialize: () => {
        throw new Error('Operation not supported.');
    },
    deserialize: (buffer, offset = 0) => {
        const pluginHeaderSerializer = (0, generated_1.getPluginHeaderV1AccountDataSerializer)();
        const pluginHeaderSize = pluginHeaderSerializer.fixedSize;
        if (pluginHeaderSize === null) {
            throw new Error('Invalid plugin header serializer configuration.');
        }
        // Deserialize base group data
        const [group, groupOffset] = (0, groupV1AccountData_1.getGroupV1AccountDataSerializer)().deserialize(buffer, offset);
        if (group.key !== generated_1.Key.GroupV1) {
            throw new Error(`Expected a Group account, got key: ${group.key}`);
        }
        let pluginHeader;
        let pluginRegistry;
        let pluginsList;
        let externalPluginAdaptersList;
        let finalOffset = groupOffset;
        const hasTrailingData = buffer.length > groupOffset;
        if (hasTrailingData && buffer.length < groupOffset + pluginHeaderSize) {
            throw new Error('Invalid Group account data: truncated plugin header.');
        }
        if (buffer.length >= groupOffset + pluginHeaderSize) {
            [pluginHeader] = pluginHeaderSerializer.deserialize(buffer, groupOffset);
            const pluginRegistryOffset = Number(pluginHeader.pluginRegistryOffset);
            if (!Number.isSafeInteger(pluginRegistryOffset) ||
                pluginRegistryOffset < groupOffset + pluginHeaderSize ||
                pluginRegistryOffset >= buffer.length) {
                throw new Error('Invalid Group account data: plugin registry offset is out of bounds.');
            }
            [pluginRegistry, finalOffset] =
                (0, pluginRegistryV1Data_1.getPluginRegistryV1AccountDataSerializer)().deserialize(buffer, pluginRegistryOffset);
            if (finalOffset > buffer.length) {
                throw new Error('Invalid Group account data: plugin registry exceeds buffer length.');
            }
            pluginsList = (0, plugins_1.registryRecordsToPluginsList)(pluginRegistry.registry, buffer);
            externalPluginAdaptersList =
                (0, externalPluginAdapters_1.externalRegistryRecordsToExternalPluginAdapterList)(pluginRegistry.externalRegistry, buffer);
        }
        return [
            {
                pluginHeader,
                ...group,
                ...pluginsList,
                ...externalPluginAdaptersList,
            },
            finalOffset,
        ];
    },
});
exports.getGroupV1AccountDataSerializer = getGroupV1AccountDataSerializer;
