// diff/lib/index.js
/*istanbul ignore start*/'use strict';

exports.__esModule = true;
exports.canonicalize = exports.convertChangesToXML = exports.convertChangesToDMP = exports.merge = exports.parsePatch = exports.applyPatches = exports.applyPatch = exports.createPatch = exports.createTwoFilesPatch = exports.structuredPatch = exports.diffArrays = exports.diffJson = exports.diffCss = exports.diffSentences = exports.diffTrimmedLines = exports.diffLines = exports.diffWordsWithSpace = exports.diffWords = exports.diffChars = exports.Diff = undefined;

/*istanbul ignore end*/var /*istanbul ignore start*/_base = require('./diff/base') /*istanbul ignore end*/;

/*istanbul ignore start*/var _base2 = _interopRequireDefault(_base);

/*istanbul ignore end*/var /*istanbul ignore start*/_character = require('./diff/character') /*istanbul ignore end*/;

var /*istanbul ignore start*/_word = require('./diff/word') /*istanbul ignore end*/;

var /*istanbul ignore start*/_line = require('./diff/line') /*istanbul ignore end*/;

var /*istanbul ignore start*/_sentence = require('./diff/sentence') /*istanbul ignore end*/;

var /*istanbul ignore start*/_css = require('./diff/css') /*istanbul ignore end*/;

var /*istanbul ignore start*/_json = require('./diff/json') /*istanbul ignore end*/;

var /*istanbul ignore start*/_array = require('./diff/array') /*istanbul ignore end*/;

var /*istanbul ignore start*/_apply = require('./patch/apply') /*istanbul ignore end*/;

var /*istanbul ignore start*/_parse = require('./patch/parse') /*istanbul ignore end*/;

var /*istanbul ignore start*/_merge = require('./patch/merge') /*istanbul ignore end*/;

var /*istanbul ignore start*/_create = require('./patch/create') /*istanbul ignore end*/;

var /*istanbul ignore start*/_dmp = require('./convert/dmp') /*istanbul ignore end*/;

var /*istanbul ignore start*/_xml = require('./convert/xml') /*istanbul ignore end*/;

/*istanbul ignore start*/function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

/* See LICENSE file for terms of use */

/*
 * Text diff implementation.
 *
 * This library supports the following APIS:
 * JsDiff.diffChars: Character by character diff
 * JsDiff.diffWords: Word (as defined by \b regex) diff which ignores whitespace
 * JsDiff.diffLines: Line based diff
 *
 * JsDiff.diffCss: Diff targeted at CSS content
 *
 * These methods are based on the implementation proposed in
 * "An O(ND) Difference Algorithm and its Variations" (Myers, 1986).
 * http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.4.6927
 */
exports. /*istanbul ignore end*/Diff = /*istanbul ignore start*/_base2 /*istanbul ignore end*/[/*istanbul ignore start*/'default' /*istanbul ignore end*/];
/*istanbul ignore start*/exports. /*istanbul ignore end*/diffChars = /*istanbul ignore start*/_character /*istanbul ignore end*/. /*istanbul ignore start*/diffChars;
exports. /*istanbul ignore end*/diffWords = /*istanbul ignore start*/_word /*istanbul ignore end*/. /*istanbul ignore start*/diffWords;
exports. /*istanbul ignore end*/diffWordsWithSpace = /*istanbul ignore start*/_word /*istanbul ignore end*/. /*istanbul ignore start*/diffWordsWithSpace;
exports. /*istanbul ignore end*/diffLines = /*istanbul ignore start*/_line /*istanbul ignore end*/. /*istanbul ignore start*/diffLines;
exports. /*istanbul ignore end*/diffTrimmedLines = /*istanbul ignore start*/_line /*istanbul ignore end*/. /*istanbul ignore start*/diffTrimmedLines;
exports. /*istanbul ignore end*/diffSentences = /*istanbul ignore start*/_sentence /*istanbul ignore end*/. /*istanbul ignore start*/diffSentences;
exports. /*istanbul ignore end*/diffCss = /*istanbul ignore start*/_css /*istanbul ignore end*/. /*istanbul ignore start*/diffCss;
exports. /*istanbul ignore end*/diffJson = /*istanbul ignore start*/_json /*istanbul ignore end*/. /*istanbul ignore start*/diffJson;
exports. /*istanbul ignore end*/diffArrays = /*istanbul ignore start*/_array /*istanbul ignore end*/. /*istanbul ignore start*/diffArrays;
exports. /*istanbul ignore end*/structuredPatch = /*istanbul ignore start*/_create /*istanbul ignore end*/. /*istanbul ignore start*/structuredPatch;
exports. /*istanbul ignore end*/createTwoFilesPatch = /*istanbul ignore start*/_create /*istanbul ignore end*/. /*istanbul ignore start*/createTwoFilesPatch;
exports. /*istanbul ignore end*/createPatch = /*istanbul ignore start*/_create /*istanbul ignore end*/. /*istanbul ignore start*/createPatch;
exports. /*istanbul ignore end*/applyPatch = /*istanbul ignore start*/_apply /*istanbul ignore end*/. /*istanbul ignore start*/applyPatch;
exports. /*istanbul ignore end*/applyPatches = /*istanbul ignore start*/_apply /*istanbul ignore end*/. /*istanbul ignore start*/applyPatches;
exports. /*istanbul ignore end*/parsePatch = /*istanbul ignore start*/_parse /*istanbul ignore end*/. /*istanbul ignore start*/parsePatch;
exports. /*istanbul ignore end*/merge = /*istanbul ignore start*/_merge /*istanbul ignore end*/. /*istanbul ignore start*/merge;
exports. /*istanbul ignore end*/convertChangesToDMP = /*istanbul ignore start*/_dmp /*istanbul ignore end*/. /*istanbul ignore start*/convertChangesToDMP;
exports. /*istanbul ignore end*/convertChangesToXML = /*istanbul ignore start*/_xml /*istanbul ignore end*/. /*istanbul ignore start*/convertChangesToXML;
exports. /*istanbul ignore end*/canonicalize = /*istanbul ignore start*/_json /*istanbul ignore end*/. /*istanbul ignore start*/canonicalize;
/*istanbul ignore end*/
