// @metaplex-foundation/mpl-core/dist/src/generated/programs/mplCore.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMplCoreProgramId = exports.getMplCoreProgram = exports.createMplCoreProgram = exports.MPL_CORE_PROGRAM_ID = void 0;
const errors_1 = require("../errors");
exports.MPL_CORE_PROGRAM_ID = 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d';
function createMplCoreProgram() {
    return {
        name: 'mplCore',
        publicKey: exports.MPL_CORE_PROGRAM_ID,
        getErrorFromCode(code, cause) {
            return (0, errors_1.getMplCoreErrorFromCode)(code, this, cause);
        },
        getErrorFromName(name, cause) {
            return (0, errors_1.getMplCoreErrorFromName)(name, this, cause);
        },
        isOnCluster() {
            return true;
        },
    };
}
exports.createMplCoreProgram = createMplCoreProgram;
function getMplCoreProgram(context, clusterFilter) {
    return context.programs.get('mplCore', clusterFilter);
}
exports.getMplCoreProgram = getMplCoreProgram;
function getMplCoreProgramId(context, clusterFilter) {
    return context.programs.getPublicKey('mplCore', exports.MPL_CORE_PROGRAM_ID, clusterFilter);
}
exports.getMplCoreProgramId = getMplCoreProgramId;
