// @metaplex-foundation/mpl-core/dist/src/generated/instructions/addAssetsToGroupV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.addAssetsToGroupV1 = exports.getAddAssetsToGroupV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getAddAssetsToGroupV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([['discriminator', (0, serializers_1.u8)()]], {
        description: 'AddAssetsToGroupV1InstructionData',
    }), (value) => ({ ...value, discriminator: 35 }));
}
exports.getAddAssetsToGroupV1InstructionDataSerializer = getAddAssetsToGroupV1InstructionDataSerializer;
// Instruction.
function addAssetsToGroupV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
    // Accounts.
    const resolvedAccounts = {
        group: {
            index: 0,
            isWritable: true,
            value: input.group ?? null,
        },
        payer: {
            index: 1,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 2,
            isWritable: false,
            value: input.authority ?? null,
        },
        systemProgram: {
            index: 3,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
    };
    // Default values.
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getAddAssetsToGroupV1InstructionDataSerializer().serialize({});
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.addAssetsToGroupV1 = addAssetsToGroupV1;
