// @metaplex-foundation/mpl-core/dist/src/generated/instructions/createV2.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createV2 = exports.getCreateV2InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getCreateV2InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['dataState', (0, types_1.getDataStateSerializer)()],
        ['name', (0, serializers_1.string)()],
        ['uri', (0, serializers_1.string)()],
        ['plugins', (0, serializers_1.option)((0, serializers_1.array)((0, types_1.getPluginAuthorityPairSerializer)()))],
        [
            'externalPluginAdapters',
            (0, serializers_1.option)((0, serializers_1.array)((0, types_1.getBaseExternalPluginAdapterInitInfoSerializer)())),
        ],
    ], { description: 'CreateV2InstructionData' }), (value) => ({
        ...value,
        discriminator: 20,
        dataState: value.dataState ?? types_1.DataState.AccountState,
        plugins: value.plugins ?? [],
        externalPluginAdapters: value.externalPluginAdapters ?? [],
    }));
}
exports.getCreateV2InstructionDataSerializer = getCreateV2InstructionDataSerializer;
// Instruction.
function createV2(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
    // Accounts.
    const resolvedAccounts = {
        asset: {
            index: 0,
            isWritable: true,
            value: input.asset ?? null,
        },
        collection: {
            index: 1,
            isWritable: true,
            value: input.collection ?? null,
        },
        authority: {
            index: 2,
            isWritable: false,
            value: input.authority ?? null,
        },
        payer: {
            index: 3,
            isWritable: true,
            value: input.payer ?? null,
        },
        owner: {
            index: 4,
            isWritable: false,
            value: input.owner ?? null,
        },
        updateAuthority: {
            index: 5,
            isWritable: false,
            value: input.updateAuthority ?? null,
        },
        systemProgram: {
            index: 6,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        logWrapper: {
            index: 7,
            isWritable: false,
            value: input.logWrapper ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getCreateV2InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.createV2 = createV2;
