// @metaplex-foundation/mpl-core/dist/src/generated/instructions/executeV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.executeV1 = exports.getExecuteV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const accounts_1 = require("../accounts");
const shared_1 = require("../shared");
function getExecuteV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['instructionData', (0, serializers_1.bytes)({ size: (0, serializers_1.u32)() })],
    ], { description: 'ExecuteV1InstructionData' }), (value) => ({ ...value, discriminator: 31 }));
}
exports.getExecuteV1InstructionDataSerializer = getExecuteV1InstructionDataSerializer;
// Instruction.
function executeV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
    // Accounts.
    const resolvedAccounts = {
        asset: {
            index: 0,
            isWritable: true,
            value: input.asset ?? null,
        },
        collection: {
            index: 1,
            isWritable: true,
            value: input.collection ?? null,
        },
        assetSigner: {
            index: 2,
            isWritable: false,
            value: input.assetSigner ?? null,
        },
        payer: {
            index: 3,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 4,
            isWritable: false,
            value: input.authority ?? null,
        },
        systemProgram: {
            index: 5,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        programId: {
            index: 6,
            isWritable: false,
            value: input.programId ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.assetSigner.value) {
        resolvedAccounts.assetSigner.value = (0, accounts_1.findAssetSignerPda)(context, {
            asset: (0, shared_1.expectPublicKey)(resolvedAccounts.asset.value),
        });
    }
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    if (!resolvedAccounts.programId.value) {
        resolvedAccounts.programId.value = programId;
        resolvedAccounts.programId.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getExecuteV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.executeV1 = executeV1;
