// @metaplex-foundation/mpl-core/dist/src/generated/instructions/transferV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.transferV1 = exports.getTransferV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getTransferV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['compressionProof', (0, serializers_1.option)((0, types_1.getCompressionProofSerializer)())],
    ], { description: 'TransferV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 14,
        compressionProof: value.compressionProof ?? (0, umi_1.none)(),
    }));
}
exports.getTransferV1InstructionDataSerializer = getTransferV1InstructionDataSerializer;
// Instruction.
function transferV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
    // Accounts.
    const resolvedAccounts = {
        asset: {
            index: 0,
            isWritable: true,
            value: input.asset ?? null,
        },
        collection: {
            index: 1,
            isWritable: false,
            value: input.collection ?? null,
        },
        payer: {
            index: 2,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 3,
            isWritable: false,
            value: input.authority ?? null,
        },
        newOwner: {
            index: 4,
            isWritable: false,
            value: input.newOwner ?? null,
        },
        systemProgram: {
            index: 5,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        logWrapper: {
            index: 6,
            isWritable: false,
            value: input.logWrapper ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getTransferV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.transferV1 = transferV1;
