// @metaplex-foundation/mpl-core/dist/src/generated/instructions/updateCollectionV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateCollectionV1 = exports.getUpdateCollectionV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
function getUpdateCollectionV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['newName', (0, serializers_1.option)((0, serializers_1.string)())],
        ['newUri', (0, serializers_1.option)((0, serializers_1.string)())],
    ], { description: 'UpdateCollectionV1InstructionData' }), (value) => ({
        ...value,
        discriminator: 16,
        newName: value.newName ?? (0, umi_1.none)(),
        newUri: value.newUri ?? (0, umi_1.none)(),
    }));
}
exports.getUpdateCollectionV1InstructionDataSerializer = getUpdateCollectionV1InstructionDataSerializer;
// Instruction.
function updateCollectionV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
    // Accounts.
    const resolvedAccounts = {
        collection: {
            index: 0,
            isWritable: true,
            value: input.collection ?? null,
        },
        payer: {
            index: 1,
            isWritable: true,
            value: input.payer ?? null,
        },
        authority: {
            index: 2,
            isWritable: false,
            value: input.authority ?? null,
        },
        newUpdateAuthority: {
            index: 3,
            isWritable: false,
            value: input.newUpdateAuthority ?? null,
        },
        systemProgram: {
            index: 4,
            isWritable: false,
            value: input.systemProgram ?? null,
        },
        logWrapper: {
            index: 5,
            isWritable: false,
            value: input.logWrapper ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Default values.
    if (!resolvedAccounts.payer.value) {
        resolvedAccounts.payer.value = context.payer;
    }
    if (!resolvedAccounts.systemProgram.value) {
        resolvedAccounts.systemProgram.value = context.programs.getPublicKey('splSystem', '11111111111111111111111111111111');
        resolvedAccounts.systemProgram.isWritable = false;
    }
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getUpdateCollectionV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.updateCollectionV1 = updateCollectionV1;
