// @metaplex-foundation/mpl-core/dist/src/generated/instructions/updateCollectionInfoV1.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateCollectionInfoV1 = exports.getUpdateCollectionInfoV1InstructionDataSerializer = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const serializers_1 = require("@metaplex-foundation/umi/serializers");
const shared_1 = require("../shared");
const types_1 = require("../types");
function getUpdateCollectionInfoV1InstructionDataSerializer() {
    return (0, serializers_1.mapSerializer)((0, serializers_1.struct)([
        ['discriminator', (0, serializers_1.u8)()],
        ['updateType', (0, types_1.getUpdateTypeSerializer)()],
        ['amount', (0, serializers_1.u32)()],
    ], { description: 'UpdateCollectionInfoV1InstructionData' }), (value) => ({ ...value, discriminator: 32 }));
}
exports.getUpdateCollectionInfoV1InstructionDataSerializer = getUpdateCollectionInfoV1InstructionDataSerializer;
// Instruction.
function updateCollectionInfoV1(context, input) {
    // Program ID.
    const programId = context.programs.getPublicKey('mplCore', 'CoREENxT6tW1HoK8ypY1SxRMZTcVPm7R94rH4PZNhX7d');
    // Accounts.
    const resolvedAccounts = {
        collection: {
            index: 0,
            isWritable: true,
            value: input.collection ?? null,
        },
        bubblegumSigner: {
            index: 1,
            isWritable: false,
            value: input.bubblegumSigner ?? null,
        },
    };
    // Arguments.
    const resolvedArgs = { ...input };
    // Accounts in order.
    const orderedAccounts = Object.values(resolvedAccounts).sort((a, b) => a.index - b.index);
    // Keys and Signers.
    const [keys, signers] = (0, shared_1.getAccountMetasAndSigners)(orderedAccounts, 'programId', programId);
    // Data.
    const data = getUpdateCollectionInfoV1InstructionDataSerializer().serialize(resolvedArgs);
    // Bytes Created On Chain.
    const bytesCreatedOnChain = 0;
    return (0, umi_1.transactionBuilder)([
        { instruction: { keys, programId, data }, signers, bytesCreatedOnChain },
    ]);
}
exports.updateCollectionInfoV1 = updateCollectionInfoV1;
