// @metaplex-foundation/mpl-core/dist/src/plugins/agentIdentity.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.agentIdentityManifest = exports.agentIdentityFromBase = exports.agentIdentityUpdateInfoArgsToBase = exports.agentIdentityInitInfoArgsToBase = void 0;
const lifecycleChecks_1 = require("./lifecycleChecks");
const pluginAuthority_1 = require("./pluginAuthority");
function agentIdentityInitInfoArgsToBase(a) {
    return {
        uri: a.uri,
        lifecycleChecks: (0, lifecycleChecks_1.lifecycleChecksToBase)(a.lifecycleChecks),
        initPluginAuthority: a.initPluginAuthority
            ? (0, pluginAuthority_1.pluginAuthorityToBase)(a.initPluginAuthority)
            : null,
    };
}
exports.agentIdentityInitInfoArgsToBase = agentIdentityInitInfoArgsToBase;
function agentIdentityUpdateInfoArgsToBase(a) {
    return {
        uri: a.uri ?? null,
        lifecycleChecks: a.lifecycleChecks
            ? (0, lifecycleChecks_1.lifecycleChecksToBase)(a.lifecycleChecks)
            : null,
    };
}
exports.agentIdentityUpdateInfoArgsToBase = agentIdentityUpdateInfoArgsToBase;
function agentIdentityFromBase(s, r, account) {
    return {
        ...s,
    };
}
exports.agentIdentityFromBase = agentIdentityFromBase;
exports.agentIdentityManifest = {
    type: 'AgentIdentity',
    fromBase: agentIdentityFromBase,
    initToBase: agentIdentityInitInfoArgsToBase,
    updateToBase: agentIdentityUpdateInfoArgsToBase,
};
