// @metaplex-foundation/mpl-core/dist/src/plugins/appData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.appDataManifest = exports.appDataFromBase = exports.appDataUpdateInfoArgsToBase = exports.appDataInitInfoArgsToBase = void 0;
const lib_1 = require("./lib");
const pluginAuthority_1 = require("./pluginAuthority");
function appDataInitInfoArgsToBase(d) {
    return {
        dataAuthority: (0, pluginAuthority_1.pluginAuthorityToBase)(d.dataAuthority),
        initPluginAuthority: d.initPluginAuthority
            ? (0, pluginAuthority_1.pluginAuthorityToBase)(d.initPluginAuthority)
            : null,
        schema: d.schema ?? null,
    };
}
exports.appDataInitInfoArgsToBase = appDataInitInfoArgsToBase;
function appDataUpdateInfoArgsToBase(d) {
    return {
        schema: d.schema ?? null,
    };
}
exports.appDataUpdateInfoArgsToBase = appDataUpdateInfoArgsToBase;
function appDataFromBase(s, r, account) {
    return {
        ...s,
        dataAuthority: (0, pluginAuthority_1.pluginAuthorityFromBase)(s.dataAuthority),
        data: (0, lib_1.parseExternalPluginAdapterData)(s, r, account),
    };
}
exports.appDataFromBase = appDataFromBase;
exports.appDataManifest = {
    type: 'AppData',
    fromBase: appDataFromBase,
    initToBase: appDataInitInfoArgsToBase,
    updateToBase: appDataUpdateInfoArgsToBase,
};
