// @metaplex-foundation/mpl-core/dist/src/plugins/dataSection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dataSectionManifest = exports.dataSectionFromBase = exports.dataSectionUpdateInfoArgsToBase = exports.dataSectionInitInfoArgsToBase = void 0;
const lib_1 = require("./lib");
const linkedDataKey_1 = require("./linkedDataKey");
const pluginAuthority_1 = require("./pluginAuthority");
function dataSectionInitInfoArgsToBase(d) {
    return {
        parentKey: (0, linkedDataKey_1.linkedDataKeyToBase)(d.parentKey),
        schema: d.schema,
    };
}
exports.dataSectionInitInfoArgsToBase = dataSectionInitInfoArgsToBase;
function dataSectionUpdateInfoArgsToBase(d) {
    // You can't update the data section directly
    return {};
}
exports.dataSectionUpdateInfoArgsToBase = dataSectionUpdateInfoArgsToBase;
function dataSectionFromBase(s, r, account) {
    return {
        ...s,
        parentKey: (0, linkedDataKey_1.linkedDataKeyFromBase)(s.parentKey),
        dataAuthority: s.parentKey.__kind !== 'LinkedLifecycleHook'
            ? (0, pluginAuthority_1.pluginAuthorityFromBase)(s.parentKey.fields[0])
            : undefined,
        data: (0, lib_1.parseExternalPluginAdapterData)(s, r, account),
    };
}
exports.dataSectionFromBase = dataSectionFromBase;
exports.dataSectionManifest = {
    type: 'DataSection',
    fromBase: dataSectionFromBase,
    initToBase: dataSectionInitInfoArgsToBase,
    updateToBase: dataSectionUpdateInfoArgsToBase,
};
