// @metaplex-foundation/mpl-core/dist/src/plugins/externalPluginAdapters.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findExtraAccounts = exports.createExternalPluginAdapterUpdateInfo = exports.createExternalPluginAdapterInitInfo = exports.isExternalPluginAdapterType = exports.externalRegistryRecordsToExternalPluginAdapterList = exports.externalPluginAdapterManifests = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const _1 = require(".");
const generated_1 = require("../generated");
const appData_1 = require("./appData");
const lifecycleChecks_1 = require("./lifecycleChecks");
const oracle_1 = require("./oracle");
const extraAccount_1 = require("./extraAccount");
const linkedAppData_1 = require("./linkedAppData");
const dataSection_1 = require("./dataSection");
const linkedLifecycleHook_1 = require("./linkedLifecycleHook");
const agentIdentity_1 = require("./agentIdentity");
exports.externalPluginAdapterManifests = {
    LifecycleHook: _1.lifecycleHookManifest,
    Oracle: oracle_1.oracleManifest,
    AppData: appData_1.appDataManifest,
    LinkedLifecycleHook: linkedLifecycleHook_1.linkedLifecycleHookManifest,
    LinkedAppData: linkedAppData_1.linkedAppDataManifest,
    DataSection: dataSection_1.dataSectionManifest,
    AgentIdentity: agentIdentity_1.agentIdentityManifest,
};
function externalRegistryRecordsToExternalPluginAdapterList(records, accountData) {
    const result = {};
    records.forEach((record) => {
        const deserializedPlugin = (0, generated_1.getExternalPluginAdapterSerializer)().deserialize(accountData, Number(record.offset))[0];
        const mappedPlugin = {
            lifecycleChecks: record.lifecycleChecks.__option === 'Some'
                ? (0, lifecycleChecks_1.lifecycleChecksFromBase)(record.lifecycleChecks.value)
                : undefined,
            authority: (0, _1.pluginAuthorityFromBase)(record.authority),
            offset: record.offset,
        };
        if (deserializedPlugin.__kind === 'LifecycleHook') {
            if (!result.lifecycleHooks) {
                result.lifecycleHooks = [];
            }
            result.lifecycleHooks.push({
                type: 'LifecycleHook',
                dataOffset: (0, umi_1.isSome)(record.dataOffset)
                    ? record.dataOffset.value
                    : undefined,
                dataLen: (0, umi_1.isSome)(record.dataLen) ? record.dataLen.value : undefined,
                ...mappedPlugin,
                ...(0, _1.lifecycleHookFromBase)(deserializedPlugin.fields[0], record, accountData),
            });
        }
        else if (deserializedPlugin.__kind === 'AppData') {
            if (!result.appDatas) {
                result.appDatas = [];
            }
            result.appDatas.push({
                type: 'AppData',
                dataOffset: (0, umi_1.isSome)(record.dataOffset)
                    ? record.dataOffset.value
                    : undefined,
                dataLen: (0, umi_1.isSome)(record.dataLen) ? record.dataLen.value : undefined,
                ...mappedPlugin,
                ...(0, appData_1.appDataFromBase)(deserializedPlugin.fields[0], record, accountData),
            });
        }
        else if (deserializedPlugin.__kind === 'Oracle') {
            if (!result.oracles) {
                result.oracles = [];
            }
            result.oracles.push({
                type: 'Oracle',
                ...mappedPlugin,
                ...(0, oracle_1.oracleFromBase)(deserializedPlugin.fields[0], record, accountData),
            });
        }
        else if (deserializedPlugin.__kind === 'LinkedLifecycleHook') {
            if (!result.linkedLifecycleHooks) {
                result.linkedLifecycleHooks = [];
            }
            result.linkedLifecycleHooks.push({
                type: 'LinkedLifecycleHook',
                ...mappedPlugin,
                ...(0, linkedLifecycleHook_1.linkedLifecycleHookFromBase)(deserializedPlugin.fields[0], record, accountData),
            });
        }
        else if (deserializedPlugin.__kind === 'LinkedAppData') {
            if (!result.linkedAppDatas) {
                result.linkedAppDatas = [];
            }
            result.linkedAppDatas.push({
                type: 'LinkedAppData',
                ...mappedPlugin,
                ...(0, linkedAppData_1.linkedAppDataFromBase)(deserializedPlugin.fields[0], record, accountData),
            });
        }
        else if (deserializedPlugin.__kind === 'DataSection') {
            if (!result.dataSections) {
                result.dataSections = [];
            }
            result.dataSections.push({
                type: 'DataSection',
                dataOffset: (0, umi_1.isSome)(record.dataOffset)
                    ? record.dataOffset.value
                    : undefined,
                dataLen: (0, umi_1.isSome)(record.dataLen) ? record.dataLen.value : undefined,
                ...mappedPlugin,
                ...(0, dataSection_1.dataSectionFromBase)(deserializedPlugin.fields[0], record, accountData),
            });
        }
        else if (deserializedPlugin.__kind === 'AgentIdentity') {
            if (!result.agentIdentities) {
                result.agentIdentities = [];
            }
            result.agentIdentities.push({
                type: 'AgentIdentity',
                ...mappedPlugin,
                ...(0, agentIdentity_1.agentIdentityFromBase)(deserializedPlugin.fields[0], record, accountData),
            });
        }
    });
    return result;
}
exports.externalRegistryRecordsToExternalPluginAdapterList = externalRegistryRecordsToExternalPluginAdapterList;
const isExternalPluginAdapterType = (plugin) => {
    if (plugin.type === 'LifecycleHook' ||
        plugin.type === 'Oracle' ||
        plugin.type === 'AppData' ||
        plugin.type === 'LinkedLifecycleHook' ||
        plugin.type === 'DataSection' ||
        plugin.type === 'LinkedAppData' ||
        plugin.type === 'AgentIdentity') {
        return true;
    }
    return false;
};
exports.isExternalPluginAdapterType = isExternalPluginAdapterType;
function createExternalPluginAdapterInitInfo({ type, ...args }) {
    const manifest = exports.externalPluginAdapterManifests[type];
    return {
        __kind: type,
        fields: [manifest.initToBase(args)],
    };
}
exports.createExternalPluginAdapterInitInfo = createExternalPluginAdapterInitInfo;
function createExternalPluginAdapterUpdateInfo({ type, ...args }) {
    const manifest = exports.externalPluginAdapterManifests[type];
    return {
        __kind: type,
        fields: [manifest.updateToBase(args)],
    };
}
exports.createExternalPluginAdapterUpdateInfo = createExternalPluginAdapterUpdateInfo;
const findExtraAccounts = (context, lifecycle, externalPluginAdapters, inputs) => {
    const accounts = [];
    externalPluginAdapters.oracles?.forEach((oracle) => {
        if (oracle.lifecycleChecks?.[lifecycle]) {
            if (oracle.baseAddressConfig) {
                accounts.push((0, extraAccount_1.extraAccountToAccountMeta)(context, oracle.baseAddressConfig, {
                    ...inputs,
                    program: oracle.baseAddress,
                }));
            }
            else {
                accounts.push({
                    pubkey: oracle.baseAddress,
                    isSigner: false,
                    isWritable: false,
                });
            }
        }
    });
    externalPluginAdapters.lifecycleHooks?.forEach((hook) => {
        if (hook.lifecycleChecks?.[lifecycle]) {
            accounts.push({
                pubkey: hook.hookedProgram,
                isSigner: false,
                isWritable: false,
            });
            hook.extraAccounts?.forEach((extra) => {
                accounts.push((0, extraAccount_1.extraAccountToAccountMeta)(context, extra, {
                    ...inputs,
                    program: hook.hookedProgram,
                }));
            });
        }
    });
    externalPluginAdapters.linkedLifecycleHooks?.forEach((hook) => {
        if (hook.lifecycleChecks?.[lifecycle]) {
            accounts.push({
                pubkey: hook.hookedProgram,
                isSigner: false,
                isWritable: false,
            });
            hook.extraAccounts?.forEach((extra) => {
                accounts.push((0, extraAccount_1.extraAccountToAccountMeta)(context, extra, {
                    ...inputs,
                    program: hook.hookedProgram,
                }));
            });
        }
    });
    return accounts;
};
exports.findExtraAccounts = findExtraAccounts;
