// @metaplex-foundation/mpl-core/dist/src/plugins/lifecycleHook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lifecycleHookManifest = exports.lifecycleHookFromBase = exports.lifecycleHookUpdateInfoArgsToBase = exports.lifecycleHookInitInfoArgsToBase = void 0;
const extraAccount_1 = require("./extraAccount");
const lifecycleChecks_1 = require("./lifecycleChecks");
const pluginAuthority_1 = require("./pluginAuthority");
const lib_1 = require("./lib");
function lifecycleHookInitInfoArgsToBase(l) {
    return {
        extraAccounts: l.extraAccounts
            ? l.extraAccounts.map(extraAccount_1.extraAccountToBase)
            : null,
        hookedProgram: l.hookedProgram,
        initPluginAuthority: l.initPluginAuthority
            ? (0, pluginAuthority_1.pluginAuthorityToBase)(l.initPluginAuthority)
            : null,
        lifecycleChecks: (0, lifecycleChecks_1.lifecycleChecksToBase)(l.lifecycleChecks),
        schema: l.schema ?? null,
        dataAuthority: l.dataAuthority
            ? (0, pluginAuthority_1.pluginAuthorityToBase)(l.dataAuthority)
            : null,
    };
}
exports.lifecycleHookInitInfoArgsToBase = lifecycleHookInitInfoArgsToBase;
function lifecycleHookUpdateInfoArgsToBase(l) {
    return {
        lifecycleChecks: l.lifecycleChecks
            ? (0, lifecycleChecks_1.lifecycleChecksToBase)(l.lifecycleChecks)
            : null,
        extraAccounts: l.extraAccounts
            ? l.extraAccounts.map(extraAccount_1.extraAccountToBase)
            : null,
        schema: l.schema ?? null,
        // TODO update dataAuthority?
    };
}
exports.lifecycleHookUpdateInfoArgsToBase = lifecycleHookUpdateInfoArgsToBase;
function lifecycleHookFromBase(s, r, account) {
    return {
        ...s,
        extraAccounts: s.extraAccounts.__option === 'Some'
            ? s.extraAccounts.value.map(extraAccount_1.extraAccountFromBase)
            : undefined,
        data: (0, lib_1.parseExternalPluginAdapterData)(s, r, account),
        dataAuthority: s.dataAuthority.__option === 'Some'
            ? (0, pluginAuthority_1.pluginAuthorityFromBase)(s.dataAuthority.value)
            : undefined,
    };
}
exports.lifecycleHookFromBase = lifecycleHookFromBase;
exports.lifecycleHookManifest = {
    type: 'LifecycleHook',
    fromBase: lifecycleHookFromBase,
    initToBase: lifecycleHookInitInfoArgsToBase,
    updateToBase: lifecycleHookUpdateInfoArgsToBase,
};
