// @metaplex-foundation/mpl-core/dist/src/plugins/linkedLifecycleHook.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.linkedLifecycleHookManifest = exports.linkedLifecycleHookFromBase = exports.linkedLifecycleHookUpdateInfoArgsToBase = exports.linkedLifecycleHookInitInfoArgsToBase = void 0;
const extraAccount_1 = require("./extraAccount");
const lifecycleChecks_1 = require("./lifecycleChecks");
const pluginAuthority_1 = require("./pluginAuthority");
function linkedLifecycleHookInitInfoArgsToBase(l) {
    return {
        extraAccounts: l.extraAccounts
            ? l.extraAccounts.map(extraAccount_1.extraAccountToBase)
            : null,
        hookedProgram: l.hookedProgram,
        initPluginAuthority: l.initPluginAuthority
            ? (0, pluginAuthority_1.pluginAuthorityToBase)(l.initPluginAuthority)
            : null,
        lifecycleChecks: (0, lifecycleChecks_1.lifecycleChecksToBase)(l.lifecycleChecks),
        schema: l.schema ? l.schema : null,
        dataAuthority: l.dataAuthority
            ? (0, pluginAuthority_1.pluginAuthorityToBase)(l.dataAuthority)
            : null,
    };
}
exports.linkedLifecycleHookInitInfoArgsToBase = linkedLifecycleHookInitInfoArgsToBase;
function linkedLifecycleHookUpdateInfoArgsToBase(l) {
    return {
        lifecycleChecks: l.lifecycleChecks
            ? (0, lifecycleChecks_1.lifecycleChecksToBase)(l.lifecycleChecks)
            : null,
        extraAccounts: l.extraAccounts
            ? l.extraAccounts.map(extraAccount_1.extraAccountToBase)
            : null,
        schema: l.schema ?? null,
    };
}
exports.linkedLifecycleHookUpdateInfoArgsToBase = linkedLifecycleHookUpdateInfoArgsToBase;
function linkedLifecycleHookFromBase(s, r, account) {
    return {
        ...s,
        extraAccounts: s.extraAccounts.__option === 'Some'
            ? s.extraAccounts.value.map(extraAccount_1.extraAccountFromBase)
            : undefined,
        dataAuthority: s.dataAuthority.__option === 'Some'
            ? (0, pluginAuthority_1.pluginAuthorityFromBase)(s.dataAuthority.value)
            : undefined,
    };
}
exports.linkedLifecycleHookFromBase = linkedLifecycleHookFromBase;
exports.linkedLifecycleHookManifest = {
    type: 'LinkedLifecycleHook',
    fromBase: linkedLifecycleHookFromBase,
    initToBase: linkedLifecycleHookInitInfoArgsToBase,
    updateToBase: linkedLifecycleHookUpdateInfoArgsToBase,
};
