// @metaplex-foundation/mpl-core/dist/src/plugins/linkedDataKey.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.linkedDataKeyFromBase = exports.linkedDataKeyToBase = void 0;
const pluginAuthority_1 = require("./pluginAuthority");
function linkedDataKeyToBase(e) {
    switch (e.type) {
        case 'LinkedLifecycleHook':
            return {
                __kind: e.type,
                fields: [e.hookedProgram],
            };
        case 'LinkedAppData':
            return {
                __kind: e.type,
                fields: [(0, pluginAuthority_1.pluginAuthorityToBase)(e.dataAuthority)],
            };
        default:
            throw new Error('Unknown LinkedDataKey type');
    }
}
exports.linkedDataKeyToBase = linkedDataKeyToBase;
function linkedDataKeyFromBase(e) {
    switch (e.__kind) {
        case 'LinkedLifecycleHook':
            return {
                type: e.__kind,
                hookedProgram: e.fields[0],
            };
        case 'LinkedAppData':
            return {
                type: e.__kind,
                dataAuthority: (0, pluginAuthority_1.pluginAuthorityFromBase)(e.fields[0]),
            };
        default:
            throw new Error('Unknown LinkedDataKey type');
    }
}
exports.linkedDataKeyFromBase = linkedDataKeyFromBase;
