// @metaplex-foundation/mpl-core/dist/src/plugins/linkedAppData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.linkedAppDataManifest = exports.linkedAppDataFromBase = exports.linkedAppDataUpdateInfoArgsToBase = exports.linkedAppDataInitInfoArgsToBase = void 0;
const pluginAuthority_1 = require("./pluginAuthority");
function linkedAppDataInitInfoArgsToBase(d) {
    return {
        dataAuthority: (0, pluginAuthority_1.pluginAuthorityToBase)(d.dataAuthority),
        initPluginAuthority: d.initPluginAuthority
            ? (0, pluginAuthority_1.pluginAuthorityToBase)(d.initPluginAuthority)
            : null,
        schema: d.schema ?? null,
    };
}
exports.linkedAppDataInitInfoArgsToBase = linkedAppDataInitInfoArgsToBase;
function linkedAppDataUpdateInfoArgsToBase(d) {
    return {
        schema: d.schema ?? null,
    };
}
exports.linkedAppDataUpdateInfoArgsToBase = linkedAppDataUpdateInfoArgsToBase;
function linkedAppDataFromBase(s, r, account) {
    return {
        ...s,
        dataAuthority: (0, pluginAuthority_1.pluginAuthorityFromBase)(s.dataAuthority),
        // plugin has no data but injected in the derivation of the asset
    };
}
exports.linkedAppDataFromBase = linkedAppDataFromBase;
exports.linkedAppDataManifest = {
    type: 'LinkedAppData',
    fromBase: linkedAppDataFromBase,
    initToBase: linkedAppDataInitInfoArgsToBase,
    updateToBase: linkedAppDataUpdateInfoArgsToBase,
};
