// @metaplex-foundation/mpl-core/dist/src/plugins/royalties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.royaltiesFromBase = exports.royaltiesToBase = exports.ruleSetToBase = void 0;
function ruleSetToBase(r) {
    const base = r;
    if (base.__kind) {
        return r;
    }
    const ruleSet = r;
    if (ruleSet.type === 'ProgramAllowList' ||
        ruleSet.type === 'ProgramDenyList') {
        return {
            __kind: ruleSet.type,
            fields: [ruleSet.addresses],
        };
    }
    return { __kind: ruleSet.type };
}
exports.ruleSetToBase = ruleSetToBase;
function royaltiesToBase(r) {
    return {
        ...r,
        ruleSet: ruleSetToBase(r.ruleSet),
    };
}
exports.royaltiesToBase = royaltiesToBase;
function royaltiesFromBase(r) {
    let ruleSet;
    if (r.ruleSet.__kind === 'ProgramAllowList') {
        ruleSet = {
            ...r.ruleSet,
            type: 'ProgramAllowList',
            addresses: r.ruleSet.fields[0],
        };
    }
    else if (r.ruleSet.__kind === 'ProgramDenyList') {
        ruleSet = {
            ...r.ruleSet,
            type: 'ProgramDenyList',
            addresses: r.ruleSet.fields[0],
        };
    }
    else {
        ruleSet = {
            ...r.ruleSet,
            type: r.ruleSet.__kind,
        };
    }
    return {
        ...r,
        ruleSet,
    };
}
exports.royaltiesFromBase = royaltiesFromBase;
