// @metaplex-foundation/mpl-core/dist/src/instructions/execute.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = void 0;
const umi_1 = require("@metaplex-foundation/umi");
const generated_1 = require("../generated");
const execute = (context, args) => executeCommon(context, args);
exports.execute = execute;
const executeCommon = (context, args) => {
    // Create a new builder to store the translated Execute instructions.
    let executeBuilder = new umi_1.TransactionBuilder();
    // Caller-level signers provided via args.signers (used for Instruction[] path).
    const callerSigners = [...(args.signers || [])];
    let builder = new umi_1.TransactionBuilder();
    // Duck-type check instead of instanceof to avoid cross-realm/bundler breakage
    // when multiple copies of the umi package are resolved.
    if ('getInstructions' in args.instructions) {
        builder = args.instructions;
    }
    else if (Array.isArray(args.instructions)) {
        args.instructions.forEach((instruction) => {
            const ixSigners = [];
            instruction.keys
                .filter((key) => key.isSigner)
                .forEach((key) => {
                const signer = callerSigners.find((signerKey) => signerKey.publicKey === key.pubkey);
                if (signer) {
                    ixSigners.push(signer);
                }
            });
            builder = builder.add({
                instruction,
                signers: ixSigners,
                bytesCreatedOnChain: 0,
            });
        });
    }
    else {
        throw new Error('No builder or instructions provided');
    }
    // eslint-disable-next-line no-restricted-syntax
    for (const ix of builder.items) {
        const [assetSigner] = (0, generated_1.findAssetSignerPda)(context, {
            asset: args.asset.publicKey,
        });
        let baseBuilder = (0, generated_1.executeV1)(context, {
            ...args,
            asset: args.asset.publicKey,
            collection: args.collection?.publicKey,
            assetSigner,
            // Forward the programID of the instruction being executed.
            programId: ix.instruction.programId,
            // Forward the data of the instruction being executed.
            instructionData: ix.instruction.data,
        });
        // If an executionDelegateRecord is provided, prepend it as the first
        // remaining account so the on-chain program can read it during validation.
        if (args.executionDelegateRecord) {
            baseBuilder = baseBuilder.addRemainingAccounts([
                {
                    pubkey: (0, umi_1.publicKey)(args.executionDelegateRecord),
                    isSigner: false,
                    isWritable: false,
                },
            ]);
        }
        executeBuilder = executeBuilder.add(baseBuilder
            // Add the instruction keys as remaining accounts.
            .addRemainingAccounts(ix.instruction.keys.map((key) => {
            // If the key is the asset signer, then we don't want to add it to the Execute instruction
            // as a signer because it only gets signed in the CPI.
            if (key.pubkey === assetSigner) {
                return {
                    pubkey: key.pubkey,
                    isSigner: false,
                    isWritable: key.isWritable,
                };
            }
            return key;
        })));
        // Capture the builder items so they can be modified.
        const executeBuilderItems = executeBuilder.items;
        // Merge caller-level signers with this item's signers, filtering out the asset signer.
        const itemSigners = [...callerSigners, ...ix.signers].filter((signer) => signer.publicKey !== assetSigner);
        executeBuilderItems[executeBuilderItems.length - 1].signers.push(...itemSigners);
        // Set the modified builder items.
        executeBuilder = executeBuilder.setItems(executeBuilderItems);
    }
    return executeBuilder;
};
