// @metaplex-foundation/mpl-core/dist/src/generated/types/externalPluginAdapterSchema.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExternalPluginAdapterSchemaSerializer = exports.ExternalPluginAdapterSchema = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var ExternalPluginAdapterSchema;
(function (ExternalPluginAdapterSchema) {
    ExternalPluginAdapterSchema[ExternalPluginAdapterSchema["Binary"] = 0] = "Binary";
    ExternalPluginAdapterSchema[ExternalPluginAdapterSchema["Json"] = 1] = "Json";
    ExternalPluginAdapterSchema[ExternalPluginAdapterSchema["MsgPack"] = 2] = "MsgPack";
})(ExternalPluginAdapterSchema = exports.ExternalPluginAdapterSchema || (exports.ExternalPluginAdapterSchema = {}));
function getExternalPluginAdapterSchemaSerializer() {
    return (0, serializers_1.scalarEnum)(ExternalPluginAdapterSchema, {
        description: 'ExternalPluginAdapterSchema',
    });
}
exports.getExternalPluginAdapterSchemaSerializer = getExternalPluginAdapterSchemaSerializer;
