// @metaplex-foundation/mpl-core/dist/src/generated/types/externalValidationResult.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExternalValidationResultSerializer = exports.ExternalValidationResult = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var ExternalValidationResult;
(function (ExternalValidationResult) {
    ExternalValidationResult[ExternalValidationResult["Approved"] = 0] = "Approved";
    ExternalValidationResult[ExternalValidationResult["Rejected"] = 1] = "Rejected";
    ExternalValidationResult[ExternalValidationResult["Pass"] = 2] = "Pass";
})(ExternalValidationResult = exports.ExternalValidationResult || (exports.ExternalValidationResult = {}));
function getExternalValidationResultSerializer() {
    return (0, serializers_1.scalarEnum)(ExternalValidationResult, {
        description: 'ExternalValidationResult',
    });
}
exports.getExternalValidationResultSerializer = getExternalValidationResultSerializer;
