// @metaplex-foundation/mpl-core/dist/src/generated/types/hookableLifecycleEvent.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getHookableLifecycleEventSerializer = exports.HookableLifecycleEvent = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var HookableLifecycleEvent;
(function (HookableLifecycleEvent) {
    HookableLifecycleEvent[HookableLifecycleEvent["Create"] = 0] = "Create";
    HookableLifecycleEvent[HookableLifecycleEvent["Transfer"] = 1] = "Transfer";
    HookableLifecycleEvent[HookableLifecycleEvent["Burn"] = 2] = "Burn";
    HookableLifecycleEvent[HookableLifecycleEvent["Update"] = 3] = "Update";
    HookableLifecycleEvent[HookableLifecycleEvent["Execute"] = 4] = "Execute";
})(HookableLifecycleEvent = exports.HookableLifecycleEvent || (exports.HookableLifecycleEvent = {}));
function getHookableLifecycleEventSerializer() {
    return (0, serializers_1.scalarEnum)(HookableLifecycleEvent, {
        description: 'HookableLifecycleEvent',
    });
}
exports.getHookableLifecycleEventSerializer = getHookableLifecycleEventSerializer;
