// @metaplex-foundation/mpl-core/dist/src/generated/types/pluginType.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPluginTypeSerializer = exports.PluginType = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var PluginType;
(function (PluginType) {
    PluginType[PluginType["Royalties"] = 0] = "Royalties";
    PluginType[PluginType["FreezeDelegate"] = 1] = "FreezeDelegate";
    PluginType[PluginType["BurnDelegate"] = 2] = "BurnDelegate";
    PluginType[PluginType["TransferDelegate"] = 3] = "TransferDelegate";
    PluginType[PluginType["UpdateDelegate"] = 4] = "UpdateDelegate";
    PluginType[PluginType["PermanentFreezeDelegate"] = 5] = "PermanentFreezeDelegate";
    PluginType[PluginType["Attributes"] = 6] = "Attributes";
    PluginType[PluginType["PermanentTransferDelegate"] = 7] = "PermanentTransferDelegate";
    PluginType[PluginType["PermanentBurnDelegate"] = 8] = "PermanentBurnDelegate";
    PluginType[PluginType["Edition"] = 9] = "Edition";
    PluginType[PluginType["MasterEdition"] = 10] = "MasterEdition";
    PluginType[PluginType["AddBlocker"] = 11] = "AddBlocker";
    PluginType[PluginType["ImmutableMetadata"] = 12] = "ImmutableMetadata";
    PluginType[PluginType["VerifiedCreators"] = 13] = "VerifiedCreators";
    PluginType[PluginType["Autograph"] = 14] = "Autograph";
    PluginType[PluginType["BubblegumV2"] = 15] = "BubblegumV2";
    PluginType[PluginType["FreezeExecute"] = 16] = "FreezeExecute";
    PluginType[PluginType["PermanentFreezeExecute"] = 17] = "PermanentFreezeExecute";
    PluginType[PluginType["Groups"] = 18] = "Groups";
})(PluginType = exports.PluginType || (exports.PluginType = {}));
function getPluginTypeSerializer() {
    return (0, serializers_1.scalarEnum)(PluginType, {
        description: 'PluginType',
    });
}
exports.getPluginTypeSerializer = getPluginTypeSerializer;
