// @metaplex-foundation/mpl-core/dist/src/generated/types/relationshipKind.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRelationshipKindSerializer = exports.RelationshipKind = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var RelationshipKind;
(function (RelationshipKind) {
    RelationshipKind[RelationshipKind["Collection"] = 0] = "Collection";
    RelationshipKind[RelationshipKind["ChildGroup"] = 1] = "ChildGroup";
    RelationshipKind[RelationshipKind["ParentGroup"] = 2] = "ParentGroup";
    RelationshipKind[RelationshipKind["Asset"] = 3] = "Asset";
})(RelationshipKind = exports.RelationshipKind || (exports.RelationshipKind = {}));
function getRelationshipKindSerializer() {
    return (0, serializers_1.scalarEnum)(RelationshipKind, {
        description: 'RelationshipKind',
    });
}
exports.getRelationshipKindSerializer = getRelationshipKindSerializer;
