// @metaplex-foundation/mpl-core/dist/src/generated/types/validationResult.js
"use strict";
/**
 * This code was AUTOGENERATED using the kinobi library.
 * Please DO NOT EDIT THIS FILE, instead use visitors
 * to add features, then rerun kinobi to update it.
 *
 * @see https://github.com/metaplex-foundation/kinobi
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.getValidationResultSerializer = exports.ValidationResult = void 0;
const serializers_1 = require("@metaplex-foundation/umi/serializers");
var ValidationResult;
(function (ValidationResult) {
    ValidationResult[ValidationResult["Approved"] = 0] = "Approved";
    ValidationResult[ValidationResult["Rejected"] = 1] = "Rejected";
    ValidationResult[ValidationResult["Pass"] = 2] = "Pass";
    ValidationResult[ValidationResult["ForceApproved"] = 3] = "ForceApproved";
})(ValidationResult = exports.ValidationResult || (exports.ValidationResult = {}));
function getValidationResultSerializer() {
    return (0, serializers_1.scalarEnum)(ValidationResult, {
        description: 'ValidationResult',
    });
}
exports.getValidationResultSerializer = getValidationResultSerializer;
